File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
  let parent_p = p,
      parent_q = q;
  
  if(hasChild(parent_p, q))
    return parent_p;

  if(hasChild(parent_q, p))
    return parent_q;

  let curr = root;

  while(hasChild(curr, p) && hasChild(curr, q)) {
    if(hasChild(curr.left, p) && hasChild(curr.left, q))
      curr = curr.left;
    else if(hasChild(curr.right, p) && hasChild(curr.right, q))
      curr = curr.right;
    else
      break;
  }
  
  return curr;
  
  function hasChild(parent: TreeNode | null, child: TreeNode | null): boolean {
    if(!parent)
      return false;
    
    if(parent == child)
      return true;
    
    return (hasChild(parent.left, child) || hasChild(parent.right, child));
  }
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode P, TreeNode Q) {

        if(P.val > root.val && Q.val > root.val)
            return lowestCommonAncestor(root.right, P, Q);
        else if(P.val < root.val && Q.val < root.val)
            return lowestCommonAncestor(root.left, P, Q);
        else
            return root;
        
//         var curr = root;
        
//         while(true) {
//             if(P.val > curr.val && Q.val > curr.val)
//                 curr = curr.right;
//             else if(P.val < curr.val && Q.val < curr.val)
//                 curr = curr.left;
//             else
//                 break;
//         }
        
//         return curr;
        
//         if(root == null || root.val == P.val || root.val == Q.val)
//             return root;
        
        
//         var left = lowestCommonAncestor(root.left, P, Q);
//         var right = lowestCommonAncestor(root.right, P, Q);
        
//         if(left == null)
//             return right;
//         else if(right == null)
//             return left;
//         else 
//             return root;
    }
}